docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,560 @@
+data pointed to by the struct MUST be allocated by the producer.  This
+includes the format and metadata strings, the arrays of buffer and children
+pointers, etc.
+
+Therefore, the consumer MUST not try to interfere with the producer's
+handling of these members' lifetime.  The only way the consumer influences
+data lifetime is by calling the base ArrowArray's release callback.
+
+Released array
+''''''''''''''
+
+A released array is indicated by setting :c:member:`ArrowArray.format` to
+NULL.  Consumers SHOULD check for a NULL format string and treat it
+accordingly (probably by erroring out).  Additionally, a released array
+MAY set :c:member:`ArrowArray.release` to NULL.
+
+Release callback semantics -- for consumers
+'''''''''''''''''''''''''''''''''''''''''''
+
+Consumers MUST call an array's release callback when they won't be using
+it anymore, but they MUST not call any of its child arrays' release callbacks
+(including the optional dictionary).  The producer is responsible for releasing
+the children.
+
+Consumers MUST check that the release callback is non-NULL before calling it.
+The release callback being NULL is not an error, it should just be ignored.
+
+In any case, a consumer MUST not try to access the ArrowArray struct anymore
+after calling its release callback -- including any associated data such
+as its children.
+
+Release callback semantics -- for producers
+'''''''''''''''''''''''''''''''''''''''''''
+
+If producers need additional information for lifetime handling (for
+example, a C++ producer may want to use ``shared_ptr`` for array and
+buffer lifetime), they MUST use the :c:member:`ArrowArray.private_data`
+member to locate the required bookkeeping information.
+
+The release callback MUST not assume that the struct will be located
+at the same memory location as when it was originally produced.  The consumer
+is free to move the struct around (see "Movability").
+
+The release callback MUST walk all children arrays (including the optional
+dictionary) and call their own release callbacks.
+
+The release callback MUST free any data area directly owned by the struct
+(such as the buffers and children arrays).
+
+The release callback MUST mark the array as released, by setting
+:c:member:`ArrowArray.format` to NULL.
+
+Additionally, the release callback MUST be idempotent, which is commonly
+achieved by setting itself to NULL.
+
+Movability
+''''''''''
+
+The consumer can *move* the ``ArrowArray`` struct by bitwise copying (or
+shallow member-wise copying).  Then it MUST mark the source struct released
+(see "released array" above for how to do it) but **without** calling the
+release callback.  This ensures that only one live copy of the struct is
+active at any given time and that lifetime is correctly communicated to
+the producer.
+
+Only the parent array (such as passed between producer and consumer) may be
+moved.  It is not recommended to move the children arrays.  It should also
```

Comment:
I don't think it's a big can of worms: since moving the child would include marking that child as released, `arrow_array_deinit(&moved_child)` should be a noop

Reply:
Instead of the recommendation against moving children, I would add to the spec that valid arrays are guaranteed to have non-released children. This implies that moving a child consumes the parent as well (so the parent should be deinit'd immediately afterwards). This would mean `arrow_array_deinit()` would be the only operation which must check whether child arrays are released.

Reply:
That's why I would say the parent must be released if the child is moved.

Reply:
The problem with move'd children is that ArrowArray structs are likely to be heap allocated (and owned) by the producer. When you move them, you'd have the consumer (the one who moves) allocate new structs on the heap.

This is the can of worm, for various reason, but mostly about the opaque release function.

Reply:
Moving does not require new heap allocations

Reply:
This really seems like something we'll need to be able to do:

- move a column out of a record batch (or equivalently a StructArray)
- move chunks from a chunked array

Reply:
@bkietz That's a fair point. I'll think about changing the spec on this.